horovod/runner/launch.py

```diff
@@ -391,6 +391,18 @@ def parse_args():
                                               action=make_override_false_action(override_args),
                                               help='Explicitly disable hierarchical allgather to prevent autotuning '
                                                    'from adjusting it.')
+    
+    group_torus_allreduce = group_params.add_mutually_exclusive_group()
+    group_torus_allreduce.add_argument('--torus-allreduce',
+                                              action=make_override_true_action(override_args),
+                                              help='Perform 2D NCCL torus allreduce between workers instead of '
+                                                   'ring allreduce. Torus allreduce is the same as hierarchical allreduce '
```

Comment:
Appreciate the suggestions! I've updated the diff accordingly. Please let me know how that looks. Thanks!
I feel that as NCCL starts to provide tree allreduce starting NCCL 2.4, we might consider migrating from mixing with MPI to pure NCCL since IIUC the original idea behind using MPI for cross node allreduce was to benefit from its tree topology but I might be missing something 🧐

Reply:
Thank you for the updates! I think this looks very nice.

I'm not really aware of the detailed history behind the hierarchical allreduce in Horovod. But yes, it seems to me that it was introduced because at the time GPU-aware MPI would scale better to a large number of nodes than NCCL. And the tree allreduce, that was introduced later to NCCL, appears to have better scaling for very large numbers of GPUs, where the latency of ring allreduce becomes expensive. So you may be very right. Mixing MPI and NCCL colllectives comes with its own costs anyway.